pkg/ingester/ingester.go

```diff
@@ -241,13 +250,47 @@ func (i *Ingester) getOrCreateInstance(instanceID string) *instance {
 
 // Query the ingests for log streams matching a set of matchers.
 func (i *Ingester) Query(req *logproto.QueryRequest, queryServer logproto.Querier_QueryServer) error {
-	instanceID, err := user.ExtractOrgID(queryServer.Context())
+	// initialize stats collection for ingester queries and set grpc trailer with stats.
+	ctx := stats.NewContext(queryServer.Context())
+	defer stats.SendAsTrailer(ctx, queryServer)
+
+	instanceID, err := user.ExtractOrgID(ctx)
 	if err != nil {
 		return err
 	}
 
 	instance := i.getOrCreateInstance(instanceID)
-	return instance.Query(req, queryServer)
+	itr, err := instance.Query(ctx, req)
+	if err != nil {
+		return err
+	}
+
+	defer helpers.LogError("closing iterator", itr.Close)
+
+	if i.cfg.QueryStore {
+		start := req.Start
+		end := req.End
+		if i.cfg.QueryStoreMaxLookBackPeriod != 0 {
+			oldestStartTime := time.Now().Add(-i.cfg.QueryStoreMaxLookBackPeriod)
+			if oldestStartTime.After(req.Start) {
+				start = oldestStartTime
+			}
+		}
+
+		if start.Before(end) {
+			req.Start = start
+			req.End = end
+
+			storeItr, err := i.store.LazyQuery(ctx, logql.SelectParams{QueryRequest: req})
+			if err != nil {
+				return err
+			}
+
+			itr.Push(storeItr)
```

Comment:
Push advance the iterator.

I would create the heap at the endl have an array of `var iters []iter.EntryIterator` at the beginning. If the array has more than one item use that item otherwise use a NewHeapIterator.